File: Algorithms/QuickSort.go
package main

import "fmt"

// Partition takes an array and sorts it to have it's partition element (arr[high]) at it's sorted index and all
// less value elements to the left and all greater elements to the right
func partition(arr []int, low int, high int) int {
	// get right most element in the array
	pivot := arr[high]
	// index where our pivot (in the sorted array) lands
	pivotIndex := low - 1

	// loop through the array
	for j := low; j <= high; j++ {
		// current element is smaller than the pivot
		if arr[j] < pivot {
			// increment pivot index
			pivotIndex++
			// move the pivot index element (bigger) to the right
			// and the current element (smaller) to the left
			arr[pivotIndex], arr[j] = arr[j], arr[pivotIndex]
		}
	}
	// swap element at the pivot index with the pivot element
	arr[pivotIndex+1], arr[high] = arr[high], arr[pivotIndex+1]

	// return the pivot index
	return pivotIndex + 1
}

// Quicksort sorts elements in an array in ascending order.
func Quicksort(arr []int, low int, high int) {
	if low < high {
		// sort the pivot element (arr[high]) to it's correct location
		pi := partition(arr, low, high)
		// run quicksort on the two halves
		Quicksort(arr, low, pi-1)
		Quicksort(arr, pi+1, high)
	}
}

func main() {
	testArr := []int{4, 3, 5, 1, 8, 5}

	Quicksort(testArr, 0, len(testArr)-1)

	fmt.Printf("%v", testArr)
}

File: Algorithms/QuickSort.js
function Partition(arr) {
    // get pivot and remove from the array
    const pivot = arr.pop();
    // allocate our smaller and bigger arrays
    const smaller = []
    const bigger = []

    // iterate over our array (without the pivot) to sort into {smaller} and {bigger}
    // items smaller or equal to the pivot go into {smaller} and items bigger go into {bigger}
    arr.forEach(elem => {
        if (elem <= pivot) {
            smaller.push(elem)
        } else {
            bigger.push(elem)
        }
    })

    // return our 3 items
    return {
        pivot,
        smaller,
        bigger
    }
}

function QuickSort(arr) {
    // if our array is less than 2 items, it's sorted!
    if (arr.length < 2) {
        return arr
    }

    // else, everything with more than 2 items can be broken down
    const {
        pivot,
        smaller,
        bigger
    } = Partition(arr)

    // and our *sorted* smaller bucket, + our pivot, and our *sorted* bigger bucket will give us the whole array
    return [ ...QuickSort(smaller), pivot, ...QuickSort(bigger) ]
}

const arr = [5, 2, 1, 6, 13, -3];

console.log(QuickSort(arr))